<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Libraries-and-Datasets" data-toc-modified-id="Libraries-and-Datasets-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Libraries and Datasets</a></span><ul class="toc-item"><li><span><a href="#Provided-Data-Dictionary-for-engine_faults" data-toc-modified-id="Provided-Data-Dictionary-for-engine_faults-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Provided Data Dictionary for <code>engine_faults</code></a></span></li></ul></li><li><span><a href="#Looking-at-engine_faults" data-toc-modified-id="Looking-at-engine_faults-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Looking at <code>engine_faults</code></a></span><ul class="toc-item"><li><span><a href="#Data-Cleanups-engine_Faults--Pre-Merge" data-toc-modified-id="Data-Cleanups-engine_Faults--Pre-Merge-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Data Cleanups <code>engine_Faults</code>--Pre-Merge</a></span></li><li><span><a href="#Final-engine_faults-before-merge" data-toc-modified-id="Final-engine_faults-before-merge-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Final <code>engine_faults</code> before merge</a></span></li></ul></li><li><span><a href="#Looking-at-diagnostics" data-toc-modified-id="Looking-at-diagnostics-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Looking at <code>diagnostics</code></a></span><ul class="toc-item"><li><span><a href="#Data-Cleanups:-diagnostics" data-toc-modified-id="Data-Cleanups:-diagnostics-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Data Cleanups: <code>diagnostics</code></a></span></li><li><span><a href="#Final-diagnostics-before-merge" data-toc-modified-id="Final-diagnostics-before-merge-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Final <code>diagnostics</code> before merge</a></span></li></ul></li><li><span><a href="#Merging-engine_faults-and-diagnostics" data-toc-modified-id="Merging-engine_faults-and-diagnostics-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Merging <code>engine_faults</code> and <code>diagnostics</code></a></span><ul class="toc-item"><li><span><a href="#List-of-Columns-With-Missing-Values" data-toc-modified-id="List-of-Columns-With-Missing-Values-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>List of Columns With Missing Values</a></span></li><li><span><a href="#Additional-Data-Cleanups-Post-Merging" data-toc-modified-id="Additional-Data-Cleanups-Post-Merging-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Additional Data Cleanups Post-Merging</a></span><ul class="toc-item"><li><span><a href="#Remove-faults-where-equipmentID-has-more-than-5-characters" data-toc-modified-id="Remove-faults-where-equipmentID-has-more-than-5-characters-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Remove faults where equipmentID has more than 5 characters</a></span></li><li><span><a href="#Remove-faults-where-active-is-set-to-False" data-toc-modified-id="Remove-faults-where-active-is-set-to-False-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Remove faults where <code>active</code> is set to <code>False</code></a></span></li><li><span><a href="#Remove-faults-in-the-vicinity-of-service-location" data-toc-modified-id="Remove-faults-in-the-vicinity-of-service-location-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Remove faults in the vicinity of service location</a></span><ul class="toc-item"><li><span><a href="#Solution-1:-Use-Decimal-Degrees-Precision-of-Coordinates" data-toc-modified-id="Solution-1:-Use-Decimal-Degrees-Precision-of-Coordinates-4.2.3.1"><span class="toc-item-num">4.2.3.1&nbsp;&nbsp;</span>Solution 1: Use Decimal Degrees Precision of Coordinates</a></span></li><li><span><a href="#Solution-2:-Use-Haversine-method-(with-numpy)" data-toc-modified-id="Solution-2:-Use-Haversine-method-(with-numpy)-4.2.3.2"><span class="toc-item-num">4.2.3.2&nbsp;&nbsp;</span>Solution 2: Use Haversine method (with numpy)</a></span></li><li><span><a href="#Finally,-dropping-the-vicinity-cases" data-toc-modified-id="Finally,-dropping-the-vicinity-cases-4.2.3.3"><span class="toc-item-num">4.2.3.3&nbsp;&nbsp;</span>Finally, dropping the vicinity cases</a></span></li></ul></li><li><span><a href="#Only-focus-on-trucks-that-were-involved-in-spn-==-1569-or-spn-==-5246" data-toc-modified-id="Only-focus-on-trucks-that-were-involved-in-spn-==-1569-or-spn-==-5246-4.2.4"><span class="toc-item-num">4.2.4&nbsp;&nbsp;</span>Only focus on trucks that were involved in <code>spn == 1569</code> or <code>spn == 5246</code></a></span></li></ul></li></ul></li><li><span><a href="#Analyzing-Missing-Values" data-toc-modified-id="Analyzing-Missing-Values-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Analyzing Missing Values</a></span><ul class="toc-item"><li><span><a href="#List-of-columns-with-missing-values" data-toc-modified-id="List-of-columns-with-missing-values-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>List of columns with missing values</a></span></li></ul></li><li><span><a href="#Sample-Truck-With-Multiple-Records:-1490" data-toc-modified-id="Sample-Truck-With-Multiple-Records:-1490-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Sample Truck With Multiple Records: <code>1490</code></a></span></li><li><span><a href="#Looking-At-service_fault_codes" data-toc-modified-id="Looking-At-service_fault_codes-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Looking At <code>service_fault_codes</code></a></span></li></ul></div>

## Libraries and Datasets

Importing necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Setting up notebook options

In [2]:
pd.options.display.max_columns = 999
pd.options.display.max_rows = 999

Importing necessary datasets

In [3]:
engine_faults = pd.read_csv(
    "../data/J1939Faults.csv", 
    low_memory=False
)

In [4]:
diagnostics = pd.read_csv(
    "../data/VehicleDiagnosticOnboardData.csv", 
    low_memory=False
)

In [5]:
# # Not using the service_fault_codes
# service_fault_codes = pd.read_excel(
#     "../data/Service Fault Codes_1_0_0_167.xlsx", 
#     sheet_name="Published 14602 App A core II",
#     engine="openpyxl"
# )

In [6]:
# # This is old data: We can't use it
# vehicle_makes = pd.read_excel(
#     '../data/Vehicle_Make.xlsx', 
#     sheet_name='Sheet1',
#     engine="openpyxl"
# )

### Provided Data Dictionary for `engine_faults`

Columns | Description
:-------|:-----------
`ESS_Id` | The event subscriber service event that contained the fault
`EventTimeStamp` | When the event took place
`eventDescription` | Brief text of meaning of the code (not always present)
`actionDescription` | Never seen this filled in
`ecuSoftwareVersion` | version string from the reporting vehicle computer system
`ecuSerialNumber` | Serial number of the reporting Engine Control Module (ECM)
`ecuModel` | Model of the reporting ECM
`ecuMake` | Manufacturer of the reporting ECM
`ecuSource` | 
`spn` | Fault code being reported
`fmi` | Failure Mode associated with the Fault Code
`active` | Whether the code is being set or being removed
`activeTransitionCount` | Number of times code has been set/unset
`faultValue` | Never seen used
`EquipmentID` | Assigned truck number of the unit in question
`MCTNumber` | Communications Terminal assigned to the truck
`Latitude` | Latitude at time of event
`Longitude` | Longitude at time of event
`LocationTimeStamp` | Time latitude and longitude were obtained

## Looking at `engine_faults`

In [7]:
display(engine_faults.shape)
display(engine_faults.head())
display(engine_faults.tail())
display(engine_faults.info())

(1187335, 20)

RecordID  ESS_Id           EventTimeStamp  \
0         1  990349  2015-02-21 10:47:13.000   
1         2  990360  2015-02-21 11:34:34.000   
2         3  990364  2015-02-21 11:35:31.000   
3         4  990370  2015-02-21 11:35:33.000   
4         5  990416  2015-02-21 11:39:41.000   

                          eventDescription  actionDescription  \
0  Low (Severity Low) Engine Coolant Level                NaN   
1                                      NaN                NaN   
2      Incorrect Data Steering Wheel Angle                NaN   
3      Incorrect Data Steering Wheel Angle                NaN   
4                                      NaN                NaN   

                     ecuSoftwareVersion ecuSerialNumber              ecuModel  \
0                               unknown         unknown               unknown   
1                               unknown         unknown               unknown   
2                               unknown         unknown               unknown   
3                               unknown         unknown               unknown   
4  22281684P01*22357957P01*22362082P01*        13063430  0USA13_13_0415_2238A   

   ecuMake  ecuSource   spn  fmi  active  activeTransitionCount  faultValue  \
0  unknown          0   111   17    True                      2         NaN   
1  unknown         11   629   12    True                    127         NaN   
2  unknown         11  1807    2   False                    127         NaN   
3  unknown         11  1807    2    True                    127         NaN   
4    VOLVO          0  4364   17   False                      2         NaN   

  EquipmentID  MCTNumber   Latitude  Longitude        LocationTimeStamp  
0        1439  105354361  38.857638 -84.626851  2015-02-21 11:34:25.000  
1        1439  105354361  38.857638 -84.626851  2015-02-21 11:35:10.000  
2        1369  105336226  41.421250 -87.767361  2015-02-21 11:35:26.000  
3        1369  105336226  41.421018 -87.767361  2015-02-21 11:36:08.000  
4        1674  105427130  38.416481 -89.442638  2015-02-21 11:39:37.000

RecordID     ESS_Id           EventTimeStamp  \
1187330   1248454  123904424  2020-03-06 14:00:26.000   
1187331   1248455  123905139  2020-03-06 14:04:23.000   
1187332   1248456  123905996  2020-03-06 14:13:38.000   
1187333   1248457  123906113  2020-03-06 14:14:13.000   
1187334   1248458  123906131  2020-03-06 14:15:34.000   

                                          eventDescription  actionDescription  \
1187330             Low (Severity Low) Catalyst Tank Level                NaN   
1187331   Condition Exists Engine Protection Torque Derate                NaN   
1187332  Abnormal Rate of Change Aftertreatment 1 Intak...                NaN   
1187333         Low (Severity Medium) Engine Coolant Level                NaN   
1187334         Low (Severity Medium) Engine Coolant Level                NaN   

                                      ecuSoftwareVersion ecuSerialNumber  \
1187330  04384413*22383729*082218154102*60701732*G1*BGT*        80156139   
1187331  04358814*06099720*030816202706*09400153*G1*BDR*        79932020   
1187332  05317106*05100987*050719120655*09401585*G1*BDR*        79880653   
1187333  04384413*22544852*090619141107*60701756*G1*BGT*             NaN   
1187334  04384413*22544852*090619141107*60701756*G1*BGT*             NaN   

                  ecuModel ecuMake  ecuSource   spn  fmi  active  \
1187330  6X1u17D1500000000   CMMNS          0  1761   17   False   
1187331  6X1u13D1500000000   CMMNS          0  1569   31    True   
1187332  6X1u13D1500000000   CMMNS          0  3216   10    True   
1187333                NaN     NaN          0   111   18    True   
1187334                NaN     NaN          0   111   18   False   

         activeTransitionCount  faultValue EquipmentID  MCTNumber   Latitude  \
1187330                      3         NaN        2282  105439740  37.094768   
1187331                      5         NaN        1994  105354084  34.390740   
1187332                      1         NaN        1850  105336308  34.430370   
1187333                      8         NaN        2377  108605700  35.030925   
1187334                      8         NaN        2377  108605700  35.027314   

         Longitude        LocationTimeStamp  
1187330 -85.897407  2020-03-06 14:00:21.000  
1187331 -79.461805  2020-03-06 14:04:59.000  
1187332 -84.920509  2020-03-06 14:14:14.000  
1187333 -85.321527  2020-03-06 14:14:49.000  
1187334 -85.323472  2020-03-06 14:15:30.000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 20 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   RecordID               1187335 non-null  int64  
 1   ESS_Id                 1187335 non-null  int64  
 2   EventTimeStamp         1187335 non-null  object 
 3   eventDescription       1126490 non-null  object 
 4   actionDescription      0 non-null        float64
 5   ecuSoftwareVersion     891285 non-null   object 
 6   ecuSerialNumber        844318 non-null   object 
 7   ecuModel               1122577 non-null  object 
 8   ecuMake                1122577 non-null  object 
 9   ecuSource              1187335 non-null  int64  
 10  spn                    1187335 non-null  int64  
 11  fmi                    1187335 non-null  int64  
 12  active                 1187335 non-null  bool   
 13  activeTransitionCount  1187335 non-null  int64  
 14  faultValue        

None

### Data Cleanups `engine_Faults`--Pre-Merge

Drop these columns. There are no data in these columns (Just `NaN`)

In [8]:
engine_faults = engine_faults.drop(columns=["actionDescription", "faultValue"])

Rename all columns to format `column_name`

In [9]:
engine_faults = engine_faults.rename(columns={
    "RecordID": "record_id",
    "ESS_Id": "ess_id",
    "EventTimeStamp": "event_timestamp",
    "eventDescription": "event_description",
    "actionDescription": "action_description",
    "ecuSoftwareVersion": "ecu_software_version",
    "ecuSerialNumber": "ecu_serial_number",
    "ecuModel": "ecu_model",
    "ecuMake": "ecu_make",
    "ecuSource": "ecu_source",
    "activeTransitionCount": "active_transition_count",
    "faultValue": "fault_value",
    "EquipmentID": "equipment_id",
    "MCTNumber": "mct_number",
    "Latitude": "latitude",
    "Longitude": "longitude",
    "LocationTimeStamp": "location_timestamp"
})

### Final `engine_faults` before merge

In [10]:
display(engine_faults.shape)
display(engine_faults.head())
display(engine_faults.tail())
display(engine_faults.info())

(1187335, 18)

record_id  ess_id          event_timestamp  \
0          1  990349  2015-02-21 10:47:13.000   
1          2  990360  2015-02-21 11:34:34.000   
2          3  990364  2015-02-21 11:35:31.000   
3          4  990370  2015-02-21 11:35:33.000   
4          5  990416  2015-02-21 11:39:41.000   

                         event_description  \
0  Low (Severity Low) Engine Coolant Level   
1                                      NaN   
2      Incorrect Data Steering Wheel Angle   
3      Incorrect Data Steering Wheel Angle   
4                                      NaN   

                   ecu_software_version ecu_serial_number  \
0                               unknown           unknown   
1                               unknown           unknown   
2                               unknown           unknown   
3                               unknown           unknown   
4  22281684P01*22357957P01*22362082P01*          13063430   

              ecu_model ecu_make  ecu_source   spn  fmi  active  \
0               unknown  unknown           0   111   17    True   
1               unknown  unknown          11   629   12    True   
2               unknown  unknown          11  1807    2   False   
3               unknown  unknown          11  1807    2    True   
4  0USA13_13_0415_2238A    VOLVO           0  4364   17   False   

   active_transition_count equipment_id  mct_number   latitude  longitude  \
0                        2         1439   105354361  38.857638 -84.626851   
1                      127         1439   105354361  38.857638 -84.626851   
2                      127         1369   105336226  41.421250 -87.767361   
3                      127         1369   105336226  41.421018 -87.767361   
4                        2         1674   105427130  38.416481 -89.442638   

        location_timestamp  
0  2015-02-21 11:34:25.000  
1  2015-02-21 11:35:10.000  
2  2015-02-21 11:35:26.000  
3  2015-02-21 11:36:08.000  
4  2015-02-21 11:39:37.000

record_id     ess_id          event_timestamp  \
1187330    1248454  123904424  2020-03-06 14:00:26.000   
1187331    1248455  123905139  2020-03-06 14:04:23.000   
1187332    1248456  123905996  2020-03-06 14:13:38.000   
1187333    1248457  123906113  2020-03-06 14:14:13.000   
1187334    1248458  123906131  2020-03-06 14:15:34.000   

                                         event_description  \
1187330             Low (Severity Low) Catalyst Tank Level   
1187331   Condition Exists Engine Protection Torque Derate   
1187332  Abnormal Rate of Change Aftertreatment 1 Intak...   
1187333         Low (Severity Medium) Engine Coolant Level   
1187334         Low (Severity Medium) Engine Coolant Level   

                                    ecu_software_version ecu_serial_number  \
1187330  04384413*22383729*082218154102*60701732*G1*BGT*          80156139   
1187331  04358814*06099720*030816202706*09400153*G1*BDR*          79932020   
1187332  05317106*05100987*050719120655*09401585*G1*BDR*          79880653   
1187333  04384413*22544852*090619141107*60701756*G1*BGT*               NaN   
1187334  04384413*22544852*090619141107*60701756*G1*BGT*               NaN   

                 ecu_model ecu_make  ecu_source   spn  fmi  active  \
1187330  6X1u17D1500000000    CMMNS           0  1761   17   False   
1187331  6X1u13D1500000000    CMMNS           0  1569   31    True   
1187332  6X1u13D1500000000    CMMNS           0  3216   10    True   
1187333                NaN      NaN           0   111   18    True   
1187334                NaN      NaN           0   111   18   False   

         active_transition_count equipment_id  mct_number   latitude  \
1187330                        3         2282   105439740  37.094768   
1187331                        5         1994   105354084  34.390740   
1187332                        1         1850   105336308  34.430370   
1187333                        8         2377   108605700  35.030925   
1187334                        8         2377   108605700  35.027314   

         longitude       location_timestamp  
1187330 -85.897407  2020-03-06 14:00:21.000  
1187331 -79.461805  2020-03-06 14:04:59.000  
1187332 -84.920509  2020-03-06 14:14:14.000  
1187333 -85.321527  2020-03-06 14:14:49.000  
1187334 -85.323472  2020-03-06 14:15:30.000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 18 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   record_id                1187335 non-null  int64  
 1   ess_id                   1187335 non-null  int64  
 2   event_timestamp          1187335 non-null  object 
 3   event_description        1126490 non-null  object 
 4   ecu_software_version     891285 non-null   object 
 5   ecu_serial_number        844318 non-null   object 
 6   ecu_model                1122577 non-null  object 
 7   ecu_make                 1122577 non-null  object 
 8   ecu_source               1187335 non-null  int64  
 9   spn                      1187335 non-null  int64  
 10  fmi                      1187335 non-null  int64  
 11  active                   1187335 non-null  bool   
 12  active_transition_count  1187335 non-null  int64  
 13  equipment_id             1187335 non-null 

None

## Looking at `diagnostics`

In [11]:
display(diagnostics.shape)
display(diagnostics.head(5))
display(diagnostics.tail(5))
display(diagnostics.info())

(12821626, 4)

Id                  Name     Value  FaultId
0   1             IgnStatus     False        1
1   2     EngineOilPressure         0        1
2   3  EngineOilTemperature  96.74375        1
3   4    TurboBoostPressure         0        1
4   5            EngineLoad        11        1

Id                      Name        Value  FaultId
12821621  12864020  EngineCoolantTemperature        181.4  1248457
12821622  12864021              ParkingBrake        False  1248457
12821623  12864022    SwitchedBatteryVoltage         14.1  1248457
12821624  12864023               DistanceLtd  28606.65625  1248457
12821625  12864024                LampStatus         1023  1248458

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12821626 entries, 0 to 12821625
Data columns (total 4 columns):
 #   Column   Dtype 
---  ------   ----- 
 0   Id       int64 
 1   Name     object
 2   Value    object
 3   FaultId  int64 
dtypes: int64(2), object(2)
memory usage: 391.3+ MB


None

### Data Cleanups: `diagnostics`

Pivotting

In [12]:
diagnostics = diagnostics.pivot(index = 'FaultId', columns='Name', values='Value').reset_index()

Rename all columns to format `column_name`

In [13]:
diagnostics = diagnostics.rename(columns={
    "FaultId": "record_id",
    "AcceleratorPedal": "accelerator_pedal",
    "BarometricPressure": "barometric_pressure",
    "CruiseControlActive": "cruise_control_active",
    "CruiseControlSetSpeed": "cruise_control_set_speed",
    "DistanceLtd": "distance_ltd",
    "EngineCoolantTemperature": "engine_coolant_temperature",
    "EngineLoad": "engine_load",
    "EngineOilPressure": "engine_oil_pressure",
    "EngineOilTemperature": "engine_oil_temperature",
    "EngineRpm": "engine_rpm",
    "EngineTimeLtd": "engine_time_ltd",
    "FuelLevel": "fuel_level",
    "FuelLtd": "fuel_ltd",
    "FuelRate": "fuel_rate",
    "FuelTemperature": "fuel_temperature",
    "IgnStatus": "ign_status",
    "IntakeManifoldTemperature": "intake_manifold_temperature",
    "LampStatus": "lamp_status",
    "ParkingBrake": "parking_brake",
    "ServiceDistance": "service_distance",
    "Speed": "speed",
    "SwitchedBatteryVoltage": "switched_battery_voltage",
    "Throttle": "throttle",
    "TurboBoostPressure": "turbo_boost_pressure"
})

### Final `diagnostics` before merge

In [14]:
# Check where we are
display(diagnostics.shape)
display(diagnostics.head())
display(diagnostics.tail())
display(diagnostics.info())

(1187335, 25)

Name  record_id accelerator_pedal barometric_pressure cruise_control_active  \
0             1                 0               14.21                 False   
1             2               NaN                 NaN                   NaN   
2             3               NaN                 NaN                   NaN   
3             4               NaN                 NaN                   NaN   
4             5               NaN                 NaN                   NaN   

Name cruise_control_set_speed distance_ltd engine_coolant_temperature  \
0                    66.48672     423178.7                      100.4   
1                         NaN          NaN                        NaN   
2                         NaN          NaN                        NaN   
3                         NaN          NaN                        NaN   
4                         NaN          NaN                        NaN   

Name engine_load engine_oil_pressure engine_oil_temperature engine_rpm  \
0             11                   0               96.74375          0   
1            NaN                 NaN                    NaN        NaN   
2            NaN                 NaN                    NaN        NaN   
3            NaN                 NaN                    NaN        NaN   
4            NaN                 NaN                    NaN        NaN   

Name engine_time_ltd fuel_level         fuel_ltd fuel_rate fuel_temperature  \
0             1632.2       43.2  12300.907429328         0              NaN   
1                NaN        NaN              NaN       NaN              NaN   
2                NaN        NaN              NaN       NaN              NaN   
3                NaN        NaN              NaN       NaN              NaN   
4                NaN        NaN              NaN       NaN              NaN   

Name ign_status intake_manifold_temperature lamp_status parking_brake  \
0         False                        78.8        1023          True   
1          True                         NaN        1279           NaN   
2           NaN                         NaN        1279           NaN   
3          True                         NaN        1279           NaN   
4           NaN                         NaN       16639           NaN   

Name service_distance speed switched_battery_voltage throttle  \
0                 NaN     0                  3276.75      NaN   
1                 NaN   NaN                      NaN      NaN   
2                 NaN   NaN                      NaN      NaN   
3                 NaN   NaN                      NaN      NaN   
4                 NaN   NaN                      NaN      NaN   

Name turbo_boost_pressure  
0                       0  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN

Name     record_id accelerator_pedal barometric_pressure  \
1187330    1248454               NaN                 NaN   
1187331    1248455               100                14.5   
1187332    1248456                 0              14.355   
1187333    1248457               1.6             14.4275   
1187334    1248458               NaN                 NaN   

Name    cruise_control_active cruise_control_set_speed distance_ltd  \
1187330                   NaN                      NaN          NaN   
1187331                  True                  64.6226     423937.9   
1187332                  True                 66.48672     465925.4   
1187333                 False                 67.72946  28606.65625   
1187334                   NaN                      NaN          NaN   

Name    engine_coolant_temperature engine_load engine_oil_pressure  \
1187330                        NaN         NaN                 NaN   
1187331                        185          51               37.12   
1187332                      186.8          62               41.18   
1187333                      181.4           0               27.26   
1187334                        NaN         NaN                 NaN   

Name    engine_oil_temperature engine_rpm engine_time_ltd fuel_level  \
1187330                    NaN        NaN             NaN        NaN   
1187331               211.4937    1310.25         10722.7       96.4   
1187332               212.8438    1340.75         9326.75        100   
1187333               221.7312     863.25          586.75       23.6   
1187334                    NaN        NaN             NaN        NaN   

Name            fuel_ltd fuel_rate fuel_temperature ign_status  \
1187330              NaN       NaN              NaN        NaN   
1187331  58979.184415546  7.647805               32       True   
1187332   65080.10587046  8.995086              NaN       True   
1187333    4042.49282573         0              NaN       True   
1187334              NaN       NaN              NaN        NaN   

Name    intake_manifold_temperature lamp_status parking_brake  \
1187330                         NaN        1023           NaN   
1187331                        98.6       18431         False   
1187332                        91.4       17407           NaN   
1187333                       100.4        1023         False   
1187334                         NaN        1023           NaN   

Name    service_distance     speed switched_battery_voltage throttle  \
1187330              NaN       NaN                      NaN      NaN   
1187331              NaN  65.01096                      NaN     73.2   
1187332              NaN   66.5741                      NaN      100   
1187333              NaN  11.84489                     14.1      100   
1187334              NaN       NaN                      NaN      NaN   

Name    turbo_boost_pressure  
1187330                  NaN  
1187331                 7.83  
1187332                 6.96  
1187333                 1.74  
1187334                  NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 25 columns):
 #   Column                       Non-Null Count    Dtype 
---  ------                       --------------    ----- 
 0   record_id                    1187335 non-null  int64 
 1   accelerator_pedal            531889 non-null   object
 2   barometric_pressure          585976 non-null   object
 3   cruise_control_active        574916 non-null   object
 4   cruise_control_set_speed     576458 non-null   object
 5   distance_ltd                 585819 non-null   object
 6   engine_coolant_temperature   586071 non-null   object
 7   engine_load                  585621 non-null   object
 8   engine_oil_pressure          586244 non-null   object
 9   engine_oil_temperature       583912 non-null   object
 10  engine_rpm                   586921 non-null   object
 11  engine_time_ltd              581366 non-null   object
 12  fuel_level                   502795 non-null   object
 1

None

## Merging `engine_faults` and `diagnostics`

In [15]:
engine_faults_diag = engine_faults.merge(
    diagnostics, 
    on = "record_id", 
    how="inner",
    suffixes = ('_faults', '_diag')
)

In [16]:
# Check where we are
display(engine_faults_diag.shape)
display(engine_faults_diag.head())
display(engine_faults_diag.tail())
display(engine_faults_diag.info())

(1187335, 42)

record_id  ess_id          event_timestamp  \
0          1  990349  2015-02-21 10:47:13.000   
1          2  990360  2015-02-21 11:34:34.000   
2          3  990364  2015-02-21 11:35:31.000   
3          4  990370  2015-02-21 11:35:33.000   
4          5  990416  2015-02-21 11:39:41.000   

                         event_description  \
0  Low (Severity Low) Engine Coolant Level   
1                                      NaN   
2      Incorrect Data Steering Wheel Angle   
3      Incorrect Data Steering Wheel Angle   
4                                      NaN   

                   ecu_software_version ecu_serial_number  \
0                               unknown           unknown   
1                               unknown           unknown   
2                               unknown           unknown   
3                               unknown           unknown   
4  22281684P01*22357957P01*22362082P01*          13063430   

              ecu_model ecu_make  ecu_source   spn  fmi  active  \
0               unknown  unknown           0   111   17    True   
1               unknown  unknown          11   629   12    True   
2               unknown  unknown          11  1807    2   False   
3               unknown  unknown          11  1807    2    True   
4  0USA13_13_0415_2238A    VOLVO           0  4364   17   False   

   active_transition_count equipment_id  mct_number   latitude  longitude  \
0                        2         1439   105354361  38.857638 -84.626851   
1                      127         1439   105354361  38.857638 -84.626851   
2                      127         1369   105336226  41.421250 -87.767361   
3                      127         1369   105336226  41.421018 -87.767361   
4                        2         1674   105427130  38.416481 -89.442638   

        location_timestamp accelerator_pedal barometric_pressure  \
0  2015-02-21 11:34:25.000                 0               14.21   
1  2015-02-21 11:35:10.000               NaN                 NaN   
2  2015-02-21 11:35:26.000               NaN                 NaN   
3  2015-02-21 11:36:08.000               NaN                 NaN   
4  2015-02-21 11:39:37.000               NaN                 NaN   

  cruise_control_active cruise_control_set_speed distance_ltd  \
0                 False                 66.48672     423178.7   
1                   NaN                      NaN          NaN   
2                   NaN                      NaN          NaN   
3                   NaN                      NaN          NaN   
4                   NaN                      NaN          NaN   

  engine_coolant_temperature engine_load engine_oil_pressure  \
0                      100.4          11                   0   
1                        NaN         NaN                 NaN   
2                        NaN         NaN                 NaN   
3                        NaN         NaN                 NaN   
4                        NaN         NaN                 NaN   

  engine_oil_temperature engine_rpm engine_time_ltd fuel_level  \
0               96.74375          0          1632.2       43.2   
1                    NaN        NaN             NaN        NaN   
2                    NaN        NaN             NaN        NaN   
3                    NaN        NaN             NaN        NaN   
4                    NaN        NaN             NaN        NaN   

          fuel_ltd fuel_rate fuel_temperature ign_status  \
0  12300.907429328         0              NaN      False   
1              NaN       NaN              NaN       True   
2              NaN       NaN              NaN        NaN   
3              NaN       NaN              NaN       True   
4              NaN       NaN              NaN        NaN   

  intake_manifold_temperature lamp_status parking_brake service_distance  \
0                        78.8        1023          True              NaN   
1                         NaN        1279           NaN              NaN   
2                         N

record_id     ess_id          event_timestamp  \
1187330    1248454  123904424  2020-03-06 14:00:26.000   
1187331    1248455  123905139  2020-03-06 14:04:23.000   
1187332    1248456  123905996  2020-03-06 14:13:38.000   
1187333    1248457  123906113  2020-03-06 14:14:13.000   
1187334    1248458  123906131  2020-03-06 14:15:34.000   

                                         event_description  \
1187330             Low (Severity Low) Catalyst Tank Level   
1187331   Condition Exists Engine Protection Torque Derate   
1187332  Abnormal Rate of Change Aftertreatment 1 Intak...   
1187333         Low (Severity Medium) Engine Coolant Level   
1187334         Low (Severity Medium) Engine Coolant Level   

                                    ecu_software_version ecu_serial_number  \
1187330  04384413*22383729*082218154102*60701732*G1*BGT*          80156139   
1187331  04358814*06099720*030816202706*09400153*G1*BDR*          79932020   
1187332  05317106*05100987*050719120655*09401585*G1*BDR*          79880653   
1187333  04384413*22544852*090619141107*60701756*G1*BGT*               NaN   
1187334  04384413*22544852*090619141107*60701756*G1*BGT*               NaN   

                 ecu_model ecu_make  ecu_source   spn  fmi  active  \
1187330  6X1u17D1500000000    CMMNS           0  1761   17   False   
1187331  6X1u13D1500000000    CMMNS           0  1569   31    True   
1187332  6X1u13D1500000000    CMMNS           0  3216   10    True   
1187333                NaN      NaN           0   111   18    True   
1187334                NaN      NaN           0   111   18   False   

         active_transition_count equipment_id  mct_number   latitude  \
1187330                        3         2282   105439740  37.094768   
1187331                        5         1994   105354084  34.390740   
1187332                        1         1850   105336308  34.430370   
1187333                        8         2377   108605700  35.030925   
1187334                        8         2377   108605700  35.027314   

         longitude       location_timestamp accelerator_pedal  \
1187330 -85.897407  2020-03-06 14:00:21.000               NaN   
1187331 -79.461805  2020-03-06 14:04:59.000               100   
1187332 -84.920509  2020-03-06 14:14:14.000                 0   
1187333 -85.321527  2020-03-06 14:14:49.000               1.6   
1187334 -85.323472  2020-03-06 14:15:30.000               NaN   

        barometric_pressure cruise_control_active cruise_control_set_speed  \
1187330                 NaN                   NaN                      NaN   
1187331                14.5                  True                  64.6226   
1187332              14.355                  True                 66.48672   
1187333             14.4275                 False                 67.72946   
1187334                 NaN                   NaN                      NaN   

        distance_ltd engine_coolant_temperature engine_load  \
1187330          NaN                        NaN         NaN   
1187331     423937.9                        185          51   
1187332     465925.4                      186.8          62   
1187333  28606.65625                      181.4           0   
1187334          NaN                        NaN         NaN   

        engine_oil_pressure engine_oil_temperature engine_rpm engine_time_ltd  \
1187330                 NaN                    NaN        NaN             NaN   
1187331               37.12               211.4937    1310.25         10722.7   
1187332               41.18               212.8438    1340.75         9326.75   
1187333               27.26               221.7312     863.25          586.75   
1187334                 NaN                    NaN        NaN             NaN   

        fuel_level         fuel_ltd fuel_rate fuel_temperature ign_status  \
1187330        NaN              NaN       NaN              NaN        NaN   
1187331       96.4  58979.184415546  7.647805               32       True   
1187332

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1187335 entries, 0 to 1187334
Data columns (total 42 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   record_id                    1187335 non-null  int64  
 1   ess_id                       1187335 non-null  int64  
 2   event_timestamp              1187335 non-null  object 
 3   event_description            1126490 non-null  object 
 4   ecu_software_version         891285 non-null   object 
 5   ecu_serial_number            844318 non-null   object 
 6   ecu_model                    1122577 non-null  object 
 7   ecu_make                     1122577 non-null  object 
 8   ecu_source                   1187335 non-null  int64  
 9   spn                          1187335 non-null  int64  
 10  fmi                          1187335 non-null  int64  
 11  active                       1187335 non-null  bool   
 12  active_transition_count      1187335 non-n

None

### List of Columns With Missing Values

- `event_description`: 
  - 60845 `na`
- `ecu_software_version`:
  - 296050 `na`
  - 298549 `unknown`
- `ecu_serial_number`:
  - 343017 `na`
  - 298549 `unknown`
- `ecu_model`:
  - 64758 `na`
  - 298549 `unknown`
- `ecu_make`:
  - 64758 `na`
  - 298549 `unknown`
  - 755 `?????`
  - 627 `????S`
  - Maybe fix the following??
    - `?MMNS` -> `CMMNS`
    - `?CAR` -> `PCAR`
    - `???R` -> `PCAR`
    - `??MNS` -> `CMMNS`
    - `?ACCR` -> `PACCR`
    - `????R` -> `PACCR`
    - `?ATON` -> `EATON`
    - `?NDWS` -> `BNDWS`
    - `??DWS` -> `BNDWS`
    - `???CR` -> `PACCR`
    - `????S` -> `unknown` (Could be CMMNS or BNDWS) -- Fill with unknown
    - `?????` -> `unknown` (Could be anything) -- Fill with unknown
- `accelerator_pedal`:
  - 655446 `na`
- `barometric_pressure`:
  - 601359 `na`
- `cruise_control_active`:
  - 612419 `na`
- `cruise_control_set_speed`:
  - 610877 `na`
- `distance_ltd`:
  - 601516 `na`
- `engine_coolant_temperature`:
  - 601264 `na`
- `engine_load`:
  - 601714 `na`
- `engine_oil_pressure`:
  - 601091 `na`
- `engine_oil_temperature`:
  - 603423 `na`
- `engine_rpm`:
  - 600414 `na`
- `engine_time_ltd`:
  - 605969 `na`
- `fuel_level`:
  - 605969 `na`
- `fuel_ltd`:
  - 602140 `na`
- `fuel_rate`:
  - 602098 `na`
- `fuel_temperature`:
  - 888225 `na`
- `ign_status`:
  - 578881 `na`
- `intake_manifold_temperature`:
  - 601044 `na`
- `parking_brake`:
  - 787363 `na`
- `service_distance`:
  - 1187120 `na`
- `speed`:
  - 603419 `na`
- `switched_battery_voltage`:
  - 1073276 `na`
- `throttle`:
  - 766832 `na`
- `turbo_boost_pressure`:
  - 603984 `na`

In [ ]:
# engine_faults_diag.groupby(["turbo_boost_pressure"]).count().reset_index()

In [ ]:
# engine_faults_diag[engine_faults_diag['location_timestamp'].str.match('\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}\.\d{3}')== True]

# for x in engine_faults_diag["turbo_boost_pressure"].unique():
#     print(x)

In [ ]:
# engine_faults_diag["turbo_boost_pressure"].nunique()
# engine_faults_diag["turbo_boost_pressure"].unique()

# engine_faults_diag[engine_faults_diag["turbo_boost_pressure"].isna()]
# engine_faults_diag[engine_faults_diag["turbo_boost_pressure"].isin(["unknown"])]
# engine_faults_diag[engine_faults_diag["turbo_boost_pressure"].isin([""])].shape

In [ ]:
# -- Target variable: Only 5246
# -- df['Derate'] = df['spn'].apply(lambda x: 1 if x == '5246' else 0)

Data Cleanups -- Post-Merge -- engine_faults

In [ ]:
# Columns To String
to_string_columns = [
    "record_id", 
    "ess_id",
    "event_description",
    "ecu_software_version",
    "ecu_serial_number",
    "ecu_model",
    "ecu_make",
    "ecu_source",
    "spn",
    "fmi",
    "equipment_id",
    "mct_number",
    #"latitude",
    #"longitude"
]

for col in to_string_columns:
    engine_faults[col] = engine_faults[col].astype("str")

In [ ]:
# Columns To Boolean
engine_faults["active"] = engine_faults["active"].astype("bool")

In [ ]:
# Columns To Integer
engine_faults["active_transition_count"] = engine_faults["active_transition_count"].astype("int64")

In [ ]:
# Columns To Float
engine_faults["latitude"] = engine_faults["latitude"].astype("float64")
engine_faults["longitude"] = engine_faults["longitude"].astype("float64")

In [ ]:
# Columns To datetime
engine_faults["event_timestamp"] = engine_faults["event_timestamp"].astype("datetime64")
engine_faults["location_timestamp"] = engine_faults["location_timestamp"].astype("datetime64")

In [ ]:
# Check where we are
display(engine_faults.shape)

Data Cleanup -- Post-Merge -- Diagnostics

In [ ]:
# Columns To String
diagnostics["fault_id"] = diagnostics["fault_id"].astype("str")
diagnostics["lamp_status"] = diagnostics["lamp_status"].astype("str")

In [ ]:
# Columns To Boolean
diagnostics["cruise_control_active"] = diagnostics["cruise_control_active"].astype("bool")
diagnostics["ign_status"] = diagnostics["ign_status"].astype("bool")
diagnostics["parking_brake"] = diagnostics["parking_brake"].astype("bool")

In [ ]:
# Columns To Float
to_float_columns = [
    "accelerator_pedal", 
    "barometric_pressure",
    "cruise_control_set_speed",
    "distance_ltd",
    "engine_coolant_temperature",
    "engine_load",
    "engine_oil_pressure",
    "engine_oil_temperature",
    "engine_rpm",
    "engine_time_ltd",
    "fuel_level",
    "fuel_ltd",
    "fuel_rate",
    "fuel_temperature",
    "intake_manifold_temperature",
    "service_distance",
    "speed",
    "switched_battery_voltage",
    "throttle",
    "turbo_boost_pressure"
]

for col in to_float_columns:
    diagnostics[col] = diagnostics[col].str.replace(',', '.')
    diagnostics[col] = diagnostics[col].astype("float64")

In [ ]:
# Change column name fault_id to record_id for consistency
diagnostics = diagnostics.rename(columns={"fault_id":"record_id"})

### Additional Data Cleanups Post-Merging

#### Remove faults where equipmentID has more than 5 characters

In [ ]:
engine_faults_diag = engine_faults_diag[engine_faults_diag["equipment_id"].map(len) <= 5]

In [ ]:
# Check where we are
display(engine_faults_diag.shape)

#### Remove faults where `active` is set to `False`

In [ ]:
engine_faults_diag = engine_faults_diag[engine_faults_diag["active"] != False]

In [ ]:
# Check where we are
display(engine_faults_diag.shape)

#### Remove faults in the vicinity of service location

The service locations:

- `(36.0666667, -86.4347222)`
- `(35.5883333, -86.4438888)`
- `(36.1950, -83.174722)`

**Coordinates Precision**

Decimal Places | Decimal Degrees | Object that can be unambiguously recognized at this scale
:-------------:|-----------------|:---------------------------------------------------------
0 | 1.0 | country or large region
1 | 0.1 | large city or district
2 | 0.01 | town or village
3 | 0.001 | neighborhood, street
4 | 0.0001 | individual street, large buildings
5 | 0.00001 | individual trees, houses
6 | 0.000001 | individual humans
7 | 0.0000001 | practical limit of commercial surveying
8 | 0.00000001 | specialized surveying (e.g. tectonic plate mapping)

##### Solution 1: Use Decimal Degrees Precision of Coordinates

In [ ]:
# # These are the cases that actually happen right *at* the service centers
# service_centers_cases = engine_faults_diag[ 
#     ((engine_faults_diag['latitude'] == 36.0666667) & (engine_faults_diag['longitude'] == -86.43472222)) |
#     ((engine_faults_diag['latitude'] == 35.5883333) & (engine_faults_diag['longitude'] == -86.4438888)) |
#     ((engine_faults_diag['latitude'] == 36.1950) & (engine_faults_diag['longitude'] == -83.174722))
# ]

# # Results Excluded: 30

In [ ]:
# display(service_centers_cases.shape)
# display(service_centers_cases.head())

Since `precision==0.0001` is the same as locating a building, we don't get enough *vicinity* using that so we will skip this precision

In [ ]:
# # These are the cases that happen *at* or within the vicinity of the centers
# # +/- .001 swing and truncate on each lat long. 
# # Precision: +/- 0.062 miles ~= +/- 328 feet ~= +/- 100 meters
# service_centers_vicinity_cases = engine_faults_diag[ 
#     ((engine_faults_diag['latitude'].between(36.065, 36.067, inclusive=True)) & (engine_faults_diag['longitude'].between(-86.435, -86.433, inclusive=True))) |
#     ((engine_faults_diag['latitude'].between(35.587, 35.589, inclusive=True)) & (engine_faults_diag['longitude'].between(-86.444, -86.442, inclusive=True))) |
#     ((engine_faults_diag['latitude'].between(36.194, 36.196, inclusive=True)) & (engine_faults_diag['longitude'].between(-83.175, -83.173, inclusive=True)))
# ]

# # Result Excluded: 36,454
# # Final Remaining: 570,882

In [ ]:
# # These are the cases that happen *at* or within the vicinity of the centers
# # +/- .01 swing and truncate on each lat long. 
# # Precision: +/- 0.62 miles ~= 1091 yard ~= +/- 1 Km
# service_centers_vicinity_cases = engine_faults_diag[ 
#     ((engine_faults_diag['latitude'].between(36.05, 36.07, inclusive=True)) & (engine_faults_diag['longitude'].between(-86.44, -86.42, inclusive=True))) |
#     ((engine_faults_diag['latitude'].between(35.57, 35.59, inclusive=True)) & (engine_faults_diag['longitude'].between(-86.45, -86.43, inclusive=True))) |
#     ((engine_faults_diag['latitude'].between(36.18, 36.20, inclusive=True)) & (engine_faults_diag['longitude'].between(-83.18, -83.16, inclusive=True)))
# ]

# # Result Excluded: 59,570
# # Final Remaining: 547,766

In [ ]:
# display(service_centers_vicinity_cases.shape)
# display(service_centers_vicinity_cases.head())

##### Solution 2: Use Haversine method (with numpy)

In [ ]:
# https://www.geeksforgeeks.org/program-distance-two-points-earth/
# Python 3 program to calculate Distance Between Two Points on Earth
from math import radians, cos, sin, asin, sqrt

def get_distance_in_km(lats1, longs1, lats2, longs2):
    """Takes a np.array of coordinates and calculate the distance between 2 coordinates in Km using the Harvesine method"""
     
    # The numpy library contains a function named np.radians() 
    # which converts an array of floats from degrees to radians.
    lats1 = np.radians(lats1)
    lats2 = np.radians(lats2)
    longs1 = np.radians(longs1)
    longs2 = np.radians(longs2)
      
    # Apply Haversine formula
    delta_longs = longs2 - longs1
    delta_lats = lats2 - lats1
    a = np.sin(delta_lats / 2)**2 + np.cos(lats1) * np.cos(lats2) * np.sin(delta_longs / 2)**2
 
    c = 2 * np.arcsin(np.sqrt(a))
    
    # Radius of earth in kilometers. Use 3956 for miles
    r = 6371
      
    # calculate the result
    return(c * r)

We can now apply this on the dataset, but we need to make some adjustments first

In [ ]:
# Add the coordinates of the service stations to the dataset
engine_faults_diag['svc_ctr_1_lat'] = 36.0666667
engine_faults_diag['svc_ctr_1_long'] = -86.43472222
engine_faults_diag['svc_ctr_2_lat'] = 35.5883333
engine_faults_diag['svc_ctr_2_long'] = -86.4438888
engine_faults_diag['svc_ctr_3_lat'] = 36.1950
engine_faults_diag['svc_ctr_3_long'] = -83.174722

In [ ]:
# Add the distance away of the truck's location from each service center 1, 2, and 3
engine_faults_diag["km_from_svcctr_1"] = get_distance_in_km(engine_faults_diag["latitude"], engine_faults_diag["longitude"], engine_faults_diag["svc_ctr_1_lat"], engine_faults_diag["svc_ctr_1_long"])
engine_faults_diag["km_from_svcctr_2"] = get_distance_in_km(engine_faults_diag["latitude"], engine_faults_diag["longitude"], engine_faults_diag["svc_ctr_2_lat"], engine_faults_diag["svc_ctr_2_long"])
engine_faults_diag["km_from_svcctr_3"] = get_distance_in_km(engine_faults_diag["latitude"], engine_faults_diag["longitude"], engine_faults_diag["svc_ctr_3_lat"], engine_faults_diag["svc_ctr_3_long"])


In [ ]:
# Let's see what those did
display(engine_faults_diag.shape)
display(engine_faults_diag.head())

Now, let's define what *Vicinity* means in terms of Km-away from the service centers

In [ ]:
# We will define "in vicinity" as within 1 Km radius
vicinity_limit = 1   # 1 Km = 0.621371 Miles

- This means that any truck within 1 Km radius of the service stations should be removed
- We need to consider all 3 service locations

In [ ]:
# These are the cases that happen *at* or within the vicinity of the centers: 1Km Radius
service_centers_vicinity_cases = engine_faults_diag[
    (engine_faults_diag["km_from_svcctr_1"] <= vicinity_limit) |
    (engine_faults_diag["km_from_svcctr_2"] <= vicinity_limit) |
    (engine_faults_diag["km_from_svcctr_3"] <= vicinity_limit)
]

# # Results Excluded: 60,232
# # Final Remaining: 547,104

In [ ]:
display(service_centers_vicinity_cases.shape)
display(service_centers_vicinity_cases.head())

##### Finally, dropping the vicinity cases

We need to drop the `service_centers_vicinity_cases` from the `engine_faults_diag`

In [ ]:
engine_faults_diag = engine_faults_diag.drop(service_centers_vicinity_cases.index, axis=0)

In [ ]:
# Check where we are
display(engine_faults_diag.shape)
display(engine_faults_diag.head())

#### Only focus on trucks that were involved in `spn == 1569` or `spn == 5246`

- It is possible that the trucks were involved in different `spn` prior to those codes
- Could that be a precursor of half- and/or full-derates?
- So instead of filtering for only these `spn`, what we want to do is focus on all the history of any trucks who were involved in these `spn`
  - They might have had different `spn` prior to these 2 `spn`

In [ ]:
# Get all the cases that are half- or full-derates
all_derates_cases = engine_faults_diag[engine_faults_diag["spn"].isin(["1569", "5246"])]

In [ ]:
display(all_derates_cases.shape)
display(all_derates_cases.head())

In [ ]:
# Which unique trucks have had a derate? We can identify this by the equipment ID
all_derates_cases["equipment_id"].unique()

In [ ]:
# Which unique trucks have had a derate? We can identify this by the equipment ID
all_derates_cases["equipment_id"].value_counts().to_frame()

We only want to focus on these trucks in the overall dataset

In [ ]:
# Filter `engine_faults_diag` for only those Equipment IDs
engine_faults_diag = engine_faults_diag[engine_faults_diag["equipment_id"].isin(all_derates_cases["equipment_id"].unique())]

In [ ]:
# Check where we are
display(engine_faults_diag.shape)
display(engine_faults_diag.head())

# # Final Remaining: 435,979

## Analyzing Missing Values

### List of columns with missing values

- `event_description`: A bunch of `nan` (This column can be dropped)
- `ecu_software_version`: A bunch of `unknown` and bad data
- `ecu_serial_number`: A bunch of `nan` and `unknown`
- `ecu_model`: A bunch of `nan` and `unknown`
- `ecu_make`: A bunch of `nan`, `unknown`, and question marks
- `ecu_source`: Existing values are 49, 0, 11, 3. Not sure if `0` is for missing values
- `spn`: 
- `active`: All `True` (Ths column could be dropped)
- `accelerator_pedal`: A bunch of `nan`
- `barometric_pressure`: A bunch of `nan`
- `cruise_control_set_speed`: A bunch of `nan`
- `distance_ltd`: A bunch of `nan`

In [ ]:
engine_faults_diag[engine_faults_diag["fmi"] == "0"]

In [ ]:
# engine_faults_diag["cruise_control_set_speed"].value_counts().sort_index()
engine_faults_diag[engine_faults_diag["distance_ltd"].isna()]
# engine_faults_diag[engine_faults_diag["cruise_control_active"].isin(["unknown"])]

## Sample Truck With Multiple Records: `1490`

Sorting by `event_timestamp` so we can see if there is a trend that leads to the derates (1569, 5246)

In [ ]:
sample_truck = engine_faults_diag[engine_faults_diag["equipment_id"] == "1490"].sort_values("event_timestamp")

In [ ]:
display(sample_truck.shape)
display(sample_truck)

In [ ]:
sample_truck = engine_faults_diag[engine_faults_diag["equipment_id"] == "1692"].sort_values("event_timestamp")

In [ ]:
display(sample_truck.shape)
display(sample_truck)

- `activeTransitionCount` - Capped out at 127 (0-127). Might not be very helpful because it is not incremental
- `Ltd` - Lifetime-to-date
- Better to pay more attention to `event_timestamp` than the `location_timestamp`
  - If you are not doing a lot of lat,long
- AcceleratorPedal - Might not have a lot of value
- EngineLoad goes up to 103
- BatteryVoltage: One value is way off

**Handling Missing Data**

- Forward Filling: `fill_na()`
- Interpolating
- Back-filling

## Looking At `service_fault_codes`

- `published_in_ces14602`: `Y` to all 7124
- `cummins_fault_code`: Each one is a unique fault code
- `revision`: Only one revision : 167
- `p_id`: Most are `Not Mapped`
- `s_id`: Most are `Not Mapped`
- `m_id`: Most are `Not Mapped`, `0`, or `128`
- `j1587_fmi`: From `0` to `14`, most are 11
- `spn`: 1915 rows
- `j1939_fmi`: From `0` to `31`, most are 31
- `j2012_pcode`: Most are not mapped
- `lamp_color`: Most are `Amber`, but also `Maintenance`, `None`, `Red`
- `lamp_device`: Most are `Warning`, but also `Maintenance`, `None`, `Stop / Shutdown`
- `cummins_description`: 5374 unique entries
- `algorithm_description`: 1879 unique entries